File: LeetCode/cpp/39.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int> &candidates, int target) {
        for (int i = 0; i < candidates.size(); ++i) {
            vector<int> temp;
            temp.push_back(candidates[i]);
            int index = 0;
            helper(candidates, target, candidates[i], temp, i);
        }
        return ret;
    }

    void helper(vector<int> &candidates, int &target, int sum, vector<int> &temp, int index) {
        if (sum == target)
            ret.push_back(temp);
        if (sum >= target)
            return;
        for (int i = index; i < candidates.size(); ++i) {
            if (sum + candidates[i] > target)
                continue;
            temp.push_back(candidates[i]);
            helper(candidates, target, sum + candidates[i], temp, i);
            temp.erase(temp.end() - 1);
        }
    }

private:
    vector<vector<int>> ret;
};

File: LeetCode/python3/39.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ret = []
        self.Backtrack(sorted(candidates), target, 0, [], ret, 0)
        return ret

    def Backtrack(self, nums, target, sum, curr, ret, j):
        if sum == target:
            ret.append(curr)
            return curr
        for i in range(j, len(nums)):
            if sum + nums[i] > target:
                break
            curr.append(nums[i])
            curr = self.Backtrack(nums, target, sum + nums[i], curr, ret, i)
            curr = curr[:-1]
        return curr
